### ISSUE TITLE
Waiting for futures and getting the result

### ISSUE THREAD

**Issue by smac89** [laverdet/node-fibers#418]
Given this code:

```typescript
const a = {b: (x: number) => x + 5};
const b = Future.wrap(a);
const c: Future<number> = b.bFuture(10);

const fiber = Fiber(() => {
    console.log("Hello");
    Future.wait(c);
    console.log("World");
});
fiber.run();
fiber.run();
```

The output is:

> Hello
> World

On the other hand, changing the fiber code to this throws an error:

```typescript
const fiber = Fiber(() => {
    console.log("Hello");
    c.wait();
    console.log("World");
});
```

Output:

> Hello
> 
> /home/chigozirim/WebstormProjects/server/node_modules/fibers/future.js:210
>                         throw new Error('Future must resolve before value is ready');
>          ^
> Error: Future must resolve before value is ready
>     at Future.get (/home/chigozirim/WebstormProjects/server/node_modules/fibers/future.js:210:10)
>     at Future.wait (/home/chigozirim/WebstormProjects/server/node_modules/fibers/future.js:449:15)
>     at fibers_1.default (/home/chigozirim/WebstormProjects/server/main.ts:13:7)

---

In the first place I don't understand why I have to call `Fiber.run` twice for the code inside the fiber to run to the end.

This code here works by calling the fiber just once:

```typescript
const meaningOfLife = Future.fromPromise(
    new Promise<number>((resolve) => resolve(42))
        .then((value: number) => `There are ${value} months in 3½ years`));

const someValueFuture = Future.fromPromise(Promise.resolve(54));

Fiber(() => {
    Future.wait([someValueFuture], meaningOfLife);
    console.log(meaningOfLife.get());
    console.log(someValueFuture.get());
}).run();
```

Output:

> There are 42 months in 3½ years
> 54

It works fine in these cases too:

```typescript
const f = (x: number) => x + 42;
const e = f.future();

Fiber(() => {
    console.log("Hello");
    console.log(e(90).wait());
    console.log("World");
}).run();
```

And

``` typescript
const f = (x: number) => x + 42;
const e = f.future();

Fiber(() => {
    console.log("Hello");
    console.log(e(90).wait());
    console.log("World");
}).run();
```
---

From my testing it seems that depending on the way the future is created (`Function.future`, `Future.wrap`, `Future.fromPromise`), it behaves differently but these behaviours are not documented anywhere.

**Comment by laverdet** [comment]
`Future.wrap` is meant to convert callback-based functions into futures. This is mentioned in the documentation. In your first example you would replace this line `const a = {b: (x: number) => x + 5};` with this `const a = {b: (x, fn) => fn(undefined, x + 5)};` for correct behavior.

I would also recommend not using `Fiber` directly and instead use `Future.task`. And at the top level you would use `Future.task(...).detach()`.
